LBRY on Kubernetes with Helm

EnigmaCurry opened the pull request:
This is a working Helm chart and HOWTO for running lbrycrd, chainquery, lbrynet, and spee.ch on Kubernetes, using the containers from this repo.

[README](https://github.com/EnigmaCurry/lbry-docker/tree/k8s-lbry/contrib/k8s-lbry)

EnigmaCurry committed:
LBRY on Kubernetes with Helm

EnigmaCurry commented:
This requires changes to the lbrycrd image outlined in #70

Leopere commented:
Would you be interested in showing a reasonable platform agnostic playthrough of this guide via a video similar to your Cloud-init?

Leopere commented:
There's quite a lot to review here so I would like to sit down with you in slack today and look over its functionality and give it a run through test.  Maybe record a video for embedding into the tutorial.

EnigmaCurry commented:
I want to refine this process a bit with a few more goals:

 * Don't install tiller on the cluster - https://rimusz.net/tillerless-helm
 * Use namespaces by default
 * Use a Makefile or frontend code to encapsulate helm and kubectl commands to make this simpler.
 * Solve the data sharing problem between spee.ch and lbrynet. (share a single volume between containers? Stream results from lbrynet into spee.ch, for spee.ch to store and cache in it's own volume?)

Leopere commented:
@EnigmaCurry I would say a shared volume is the solution for shared data between spee.ch and lbrynet.

The makefile could be a really good option however BASH is generally more available than Make is it possible to just use a bash script?

EnigmaCurry committed:
k8s-lbry: Adds bash wrapper run.sh and new configs

Squashed commit of the following:

commit ce1e033f36f0b1e41c48cfa09b4660771af0e6f3
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Fri May 17 09:04:05 2019 -0400

    Added spee.ch README and speech-lbrynet services.

commit 7aa73f3d58885c7e2ff56161861fe9d9c73b70e0
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Thu May 16 19:03:28 2019 -0400

    Fix Speech downloads dirs

commit 3137b0c82e9f83482a48b721d176794efeeb6164
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Thu May 16 17:52:55 2019 -0400

    speech-lbrynet should mount speech-data volume to store downloads.

commit 727c59800a91cedf545b1f3f767e9481455c1897
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Thu May 16 17:40:34 2019 -0400

    applying shell-check (part 1)

commit 251e2b104b7f87cc0b3e430371be7a6fdfca1d08
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Wed May 15 20:43:18 2019 -0400

    TLS / HTTPs

commit b3de768909a827a2a3b65e7ecbe1795d1ab75e7b
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Wed May 15 17:51:52 2019 -0400

    Spee.ch+lbrynet sidecar (in progress)

commit 1eb08ae6f9dd57e4bdfe9c196493f609908e4105
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Tue May 14 10:43:20 2019 -0400

    Add asciicast

commit 067e9c8325c5148fb16b3a96e8c4de7d39c8f25d
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Mon May 13 09:49:50 2019 -0400

    Updates lbrynet image

commit b85026d6488c9b00d2cb8ad86f4be626db2e9d66
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Mon May 13 08:36:40 2019 -0400

    README

commit 1f5bd113205164fcc4e279e5ef51b2dad3a8e477
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 21:40:46 2019 -0400

    fix helm app names

commit 28ea709507643a4385bd0618b1de9b69f62752eb
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 21:31:22 2019 -0400

    lbrycrd-override-snapshot

commit 0d5b2ae7743e721c08d940a15e73540ce4411886
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 21:00:24 2019 -0400

    remove set -x

commit b2d380197befe8de716ea2f09bfdf7f984d2a867
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 20:41:18 2019 -0400

    fix lbrynet-copy-wallet

commit 2dcb98dee79eacb64d9ca14b0eed6a90c249c6fe
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 20:29:59 2019 -0400

    run.sh messaging

commit b766ffc624aa165c2a987bd7cd7911e3dd0a78df
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 20:21:21 2019 -0400

    fix linkages

commit 4d07652d1af1fd741bbec6debe8462bcd9e6cf65
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 19:39:23 2019 -0400

    fix container linkage

commit 8f4bd26ac7b658d3cc1b201a6870fe889e3a39d9
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 19:26:53 2019 -0400

    run.sh help

commit ef2f0fe4b7c5b5e3aef5d14b99a3293a41e45ae5
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 18:43:43 2019 -0400

    remove old messaging

commit b8a6696b0f38724ce5f0472943233da681181a72
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 17:19:33 2019 -0400

    add setup-alias command

commit fbc5c202fe48e0f5867b899795fb4e9ab696e96f
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 16:06:04 2019 -0400

    README

commit 34410c1e1f4fd3546b76a55dc0fdb713ed8a1279
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 15:51:20 2019 -0400

    Depend on external kubectl and helm tab completion scripts.

commit a9af588c42f8bcd11266733646d7990ab722a84c
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 15:23:35 2019 -0400

    Fix tab completion script

commit c82efd5463af24118249240b738a70eb0f5e9d3d
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 15:12:21 2019 -0400

    Refactor run.sh script to k8s-lbry alias.

commit d14eb6691226788f0e08cc5b162fd29de8c1782f
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 14:44:16 2019 -0400

    Fix tab completion script.

commit 77ecaab2f934290e20da969182c7f09f71d1b819
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 14:22:25 2019 -0400

    Fix kubectl namespace during tab completion

commit 29ff6d25b0256a51a05c96a0f8f9a7d443ff58d1
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sun May 12 12:13:03 2019 -0400

    k8s-lbry: Adds tab completion

commit da961eb75ea5468ad4240bf7ad4f8a8058a0909f
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 20:54:43 2019 -0400

    README

commit 934e43be256aee490385cbf9f40115e464ba930e
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 20:23:03 2019 -0400

    Chainquery snapshot restore

commit 56029260f3f6c436847efb89a37fa206f5d7265f
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 18:13:58 2019 -0400

    lbrycrd-cli credentials

commit 03bc538174e874cbf82ec061edb184583ee293b9
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 18:07:25 2019 -0400

    Open port 9246

commit 339688bb28f9fd63bdce944c55d5d2d0c57d87be
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 17:37:15 2019 -0400

    delete old package script

commit 4cef33826bcaef72eb3acecfb83d2627bd2a1dd7
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 17:31:17 2019 -0400

    fix pod names

commit babcb688a747d58b7c8b4cf5c44f46d36307d111
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 16:47:30 2019 -0400

    more

commit 88192a8992604aa9fb7edbb41d0194b70a7a00ca
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 16:13:27 2019 -0400

    enable args

commit 01bcf85e1478b47814788cc4f1d44b5b91a0ea26
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 15:55:14 2019 -0400

    more

commit 76ab5019b85a43b04554af8da6a55d30d13dedf3
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 15:41:36 2019 -0400

    k8s-lbry: lbrynet

commit 3432c9e3a089e5789c1b30d5119cfb25d087d707
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 11:41:32 2019 -0400

    README

commit 813b6ab95b2320000405dca973492d6cba19b288
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 11:40:02 2019 -0400

    fix regex

commit d010eb5ee96fded843247e0d12980f38b910c9f1
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 11:36:47 2019 -0400

    Add git dependency

commit 092c496f9bf3e86dc1a138ac404daa582fe1632e
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 11:27:14 2019 -0400

    Fix helm

commit fa7dea7139031931ff372024ee914a60d68b6157
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 10:57:17 2019 -0400

    Make run.sh bootstrappable

commit f512878a70f70f85f2e7f5bc7c239d681846e508
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Sat May 11 10:21:35 2019 -0400

    k8s-lbry: README

commit fd77fa2527962198b204a3de852c0b891e078178
Author: Ryan McGuire <ryan@enigmacurry.com>
Date:   Wed May 8 02:50:24 2019 -0400

    k8s-lbry: Adds run.sh wrapper script and new README.

EnigmaCurry commented:
I've updated my k8s-lbry branch with the new additions outlined above.

 * There is a new run.sh wrapper script
 * Tiller is run locally from run.sh and no longer installed on the cluster
 * Everything installed to the k8s-lbry namespace
 * Speech includes its own lbrynet sidecar container (and mounts a shared data dir between the two)

Right now I'm building my own spee.ch container from https://gist.github.com/EnigmaCurry/57e9773d9b3748b248a52fcebd447fba

Things to do before merge:
 * Finish production spee.ch container 
 * Update image references in values.yaml
 * Update HELM_REPO in run.sh to lbryio official repository.
 * Publish production lbrynet v0.33.0 as this is the version that works with spee.ch right now
 * Figure out why uploads don't work in spee.ch

EnigmaCurry committed:
k8s-lbry: Adds asciinema asciicast.

EnigmaCurry commented:
* Spee.ch upstream has been upgraded to SDK 0.37.2, speech-lbrynet container updated to new version.
 * Made PR #74 for compatible spee.ch dockerfile
 * Made PR #73 for building new lbrynet container. This needs to run on 0.37.2.
 * Anonymous publishing is working, there is an error in the logs though:

```
2019-05-30T21:16:20.981Z - error: Error on /api/file/availability/479364455881011fc6306o2600/86e53ddc594fcc13e558dc0c1f94b31f4ee1fb99 
{ stack: 'TypeError: Cannot read property \'dataValues\' of null\n    at /spee.ch/server/bundle/server.js:1:638181\n    at <anonymous>',
  message: 'Cannot read property \'dataValues\' of null' }
```
 * Once the claim is accepted onto the blockchain, the image resolves correctly. 
 * Test page (temporarily online only) : https://speech.rymcg.tech/8/479364455881011fc6306o2600
 * Publishing within a channel is not working, with an error message of `no channel password provided`, although logging into the channel appears to work, and I do see the password in the log when the account was created.
 * run.sh still points to my forked helm chart values, this still needs to be switched before merged.
 * values.yaml contains docker images for my own repository that needs to be switched once we have spee.ch and lbrynet (v0.37.2) production images.
 * I added the asciicast directly to the `contrib/k8s-lbry/kick-ascii` directory. Turn on github pages in the repository settings for the `master` branch.
 * There is a helm repository created on S3 that is being used currently, this could be migrated to github pages as well. That would mean mainting a full history of tarballed helm chart releases in `contrib/k8s-lbry/repository`. Alternatively, we could just keep it as S3, and host it elsewhere. I think the latter option is a bit cleaner.

EnigmaCurry committed:
Fix asciicast run.sh link

EnigmaCurry committed:
k8s-lbry: v0.1.5

Leopere commented on contrib/k8s-lbry/README.md:
```diff
@@ -0,0 +1,862 @@
+Deploy lbrycrd, lbrynet, chainquery, mysql, and spee.ch on your Kubernetes
+cluster.
+
+[![asciicast](https://enigmacurry.github.io/lbry-docker/contrib/k8s-lbry/kick-ascii/cast/k8s-lbry.png)](https://enigmacurry.github.io/lbry-docker/contrib/k8s-lbry/kick-ascii/?cast=k8s-lbry&bg=lbry.png)
```
Ideally, this should be hosted on this repository completely.

Leopere commented on contrib/k8s-lbry/kick-ascii/README.md:
```diff
@@ -0,0 +1,47 @@
+[asciinema-player](https://github.com/asciinema/asciinema-player). Self host
+your own ASCII theatre on the web.
+
+[![asciicast](http://enigmacurry.github.io/kick-ascii/cast/index.jpg)](http://enigmacurry.github.io/kick-ascii/?cast=index&bg=matrix.jpg)
```
Can these things be moved into a directory within this repository?

EnigmaCurry replied:
I copied everything from https://github.com/enigmacurry/kick-ascii into this directory. This is just the upstream README. I can actually just delete the file, its not needed..

Leopere commented on contrib/k8s-lbry/README.md:
```diff
@@ -0,0 +1,862 @@
+
+<!-- tocstop -->
+
+## Requirements
```
Just for the sake of ramping up adoption of this method, we should definitely attempt to have a proper screencast captured and hosted on LBRY for this.

Nothing worse than making the wrong assumptions we should really aim to have a full zero to hero guide for everything.

EnigmaCurry replied:
asciicast links should resolve to lbryio.github.io now.

Leopere commented on contrib/k8s-lbry/README.md:
```diff
@@ -0,0 +1,862 @@
+If everything is working, you should see a list of one or more nodes running and
+showing `STATUS=Ready`
+
+## Security Notice
```
We need to also refer to vulnerability reporting and any responsible disclosure stuff in a security notice.

EnigmaCurry replied:
added!

Leopere commented on contrib/k8s-lbry/README.md:
````diff
@@ -0,0 +1,862 @@
+
+```
+curl -Lo run.sh https://raw.githubusercontent.com/EnigmaCurry/lbry-docker/k8s-lbry/contrib/k8s-lbry/run.sh
+
````
Possibly put a comment here mentioning that people should always review the contents of a curl'd script for secure operations involving money.

EnigmaCurry replied:
added

Leopere commented on contrib/k8s-lbry/run.sh:
```diff
@@ -0,0 +1,661 @@
+
+}
+
+setup-alias() {
```
This should verify that the user's shell is BASH if you're planning on adding this.

EnigmaCurry replied:
Now it does. thank you.

Leopere commented on contrib/k8s-lbry/README.md:
````diff
@@ -0,0 +1,862 @@
+k8s-lbry kubectl get svc nginx-ingress-controller -o wide
+```
+
+If you find a hostname instead of an IP address, this means your load balancer
````
Could we not replace part of this with a script that would just use the OS's included `dig` installation to do this?  If that sounds like a proper pain in the ass nevermind but just sounds like a step we can do for them potentially.

EnigmaCurry replied:
This whole paragraph is an implementation detail that only affects AWS users (I'm not sure if it affectrs *all* AWS users, or if it was just my config). On DigitalOcean its a non-issue. I will add a comment they can use `dig` themselves. Scripting this away might make this harder to debug I think, plus end-users might have a better idea of how to route via two ip address than me?

EnigmaCurry replied:
added note to use dig if available, but otherwise leaving this as-is.

Leopere commented on contrib/k8s-lbry/README.md:
```diff
@@ -0,0 +1,862 @@
+ 
+   * mysql for storing the speech database.
+
+Edit `values-dev.yaml`. 
```
Can you update references to these files with URL's referencing their location in the repository?  Just might help people know where the defaults are.

EnigmaCurry replied:
values-dev.yaml does not exist in the repository, only values.yaml does. values-dev is created by run.sh when the user runs setup.

Leopere commented on contrib/k8s-lbry/README.md:
````diff
@@ -0,0 +1,862 @@
+Copy the wallet in case the volume gets destroyed:
+
+```
+k8s-lbry speech-lbrynet-copy-wallet /tmp/k8s-lbry-speech-lbrynet-wallet-backup.json
````
Could we potentially `cat` and pipe this into `apt-get install qrencode` on the host somehow so the user could scan it into a wallet app or something for backup?

EnigmaCurry replied:
I don't think I want to introduce a new dependency, but adding instructions to do this is a great idea!

EnigmaCurry replied:
actually if the dependency is only for one subcommand, maybe its ok. Lemme think about this. Its a grand idea.

EnigmaCurry replied:
OK after thinking about it, I think this is the wrong place to implement this. I feel like lbrynet itself should have this feature. Otherwise someone might look at the lbrynet docs, and not know what to do with this qrcode.

Leopere commented on contrib/k8s-lbry/README.md:
```diff
@@ -0,0 +1,862 @@
+# LBRY on Kubernetes with Helm
```
One of the things that I would love for the documentation is to avoid making assumptions of knowledge in general. So when a user jumps into wanting to host their own K8's based instance of this we should assume that the user will want links back to the official documentation for various processes.

EnigmaCurry replied:
added

Leopere commented on contrib/k8s-lbry/README.md:
````diff
@@ -0,0 +1,862 @@
+Upgrade nginx-ingress, turning on HTTPs support:
+
+```
+NGINX_ENABLE_HTTPS=true k8s-lbry upgrade-nginx-ingress
````
Is this a whole env-var? Looks like a combination between a variable and a command.  This should be clearer.

EnigmaCurry replied:
how about this?

```
(
  NGINX_ENABLE_HTTPS=true 
  k8s-lbry upgrade-nginx-ingress
)
```
Only problem with that though is if the user thinks they don't need the parentheses. The env var is only supposed to be temporary for that one command.

EnigmaCurry replied:
I'll just leave a note to run it with the parens.

Leopere reviewed (changes requested):
Looks like the Readme is 97% of the way there let us get it the last 3% and I'll push this into the repository!

EnigmaCurry committed:
k8s-lbry README updates.

EnigmaCurry committed:
k8s-lbry: Add warning in setup-alias if the user is not running bash.

Leopere commented:
Just waiting on @tiger5226 to enable the github pages for this repo so I can rebase and merge this at this point.

tiger5226 commented:
its enabled
